pretext/codechat.py

```diff
@@ -0,0 +1,110 @@
+# ***********************************************************
+# codechat.py -- Compute mapping used by the CodeChat System
+# ***********************************************************
+# This function computes a mapping between source file names and the resulting HTML files built from them by PreTeXt. The `CodeChat System <https://codechat-system.readthedocs.io>`_ then uses this mapping to synchronize between a source file and its resulting rendered form.
+#
+#
+# Imports
+# =======
+# These are listed in the order prescribed by `PEP 8
+# <http://www.python.org/dev/peps/pep-0008/#imports>`_.
+#
+# Standard library
+# ----------------
+import collections  # defaultdict
+import glob  # glob
+import json  # dumps
+import pathlib  # Path
+import sys  # platform
+import urllib.parse  # urlparse
+
+# Third-party imports
+# -------------------
+# We assume a previous call to ``xsltproc`` has already verified that lxml is installed.
+import lxml.etree as ET
+import lxml.ElementInclude
+
+# Local application imports
+# -------------------------
+# None.
+
+
+# Mapping
+# =======
+# Build a mapping between XML IDs and the resulting generated HTML files. The goal: map from source files to the resulting HTML files produced by the pretext build. The data structure is:
+#
+# .. code::
+#   :number-lines:
+#
+#   path_to_xml_id: Dict[
+#       # A path to the source file
+#       str,
+#       # A list of XML IDs in this source file which produce HTML files.
+#       List[str]
+#   ]
+#
+# This allows a single source file to produce multiple HTML files, as well as supporting a one-to-one relationship. The list captures the order of appearance of the XML IDs in the tree -- element 0 is the first XML ID, etc.
+def map_path_to_xml_id(
```

Comment:
I'd appreciate a minimal test added to the suite so I'll know if I break this later by accident. (-:

Reply:
Will do!